lib/Core/Executor.cpp

```diff
@@ -905,14 +905,21 @@ Executor::fork(ExecutionState &current, ref<Expr> condition, bool isInternal) {
       }
     }
 
+    // We don't add a constraint explicitly yet
```

Comment:
@ccadar Yes, that bugged me as well. The problem here is, that we don't add the condition explicitly as the solver already proofed that the query is valid (or its negation) and we spare it as an optimization (good question is, if it is really one). Still we could add the condition. 

The code example is interesting as it is a tight endless loop with a symbolic. In the context of BFS this means that if the state traversing that loop ends up in the queue first, we won't have any opportunity to add a constraint anymore. Therefore the state will be at the beginning of the BFS queue forever (as it happens with the current implementation).
With the flag, we break that tie.

From an abstract point of view, we mimic the behavior of an SE engine, which always forks, adds the appropriate condition to the forked states, and after that checks their validity.
With the flag, it indicates that the state has been forked and the condition been added; even if we don't do this as an optimization.